java/arcs/core/storage/StorageProxy.kt

```diff
@@ -0,0 +1,221 @@
+     * Disconnects a handle so it no longer receives callbacks.
+     */
+    suspend fun deregisterHandle(handle: Handle<Data, Op, T>) {
+        mutex.withLock {
+            handles.remove(handle)
+        }
+    }
+
+    /**
+     * Apply a CRDT operation to the [CrdtModel] that this [StorageProxy] manages, notifies read
+     * handles, and forwards the write to the [Store].
+     */
+    suspend fun applyOp(op: Op): Boolean {
+        mutex.withLock {
+            if (!crdt.applyOperation(op)) {
+                return false
+            }
+            val msg = ProxyMessage.Operations<Data, Op, T>(listOf(op), null)
+            store.onProxyMessage(msg)
+            notifyUpdate(op)
+            return true
+        }
+    }
+
+    /**
+     * Return the current local version of the model, as well as the current associated version
+     * map for the data.
+     * Suspends until it has a synchronized view of the data.
+     */
+    suspend fun getParticleView(): ValueAndVersion<T> {
+        val future = CompletableDeferred<ValueAndVersion<T>>()
+        mutex.withLock {
+            if (synchronized) {
+                return ValueAndVersion(crdt.consumerView, crdt.versionMap)
+            }
+            waitingSyncs.add(future)
+        }
+        return future.await()
+    }
+
+    /**
+     * Applies messages from a [Store].
+     */
+    suspend fun onMessage(message: ProxyMessage<Data, Op, T>): Boolean {
+        mutex.withLock {
+            when (message) {
+                is ProxyMessage.ModelUpdate -> {
+                    this.crdt.merge(message.model);
+                    this.setSynchronized();
+                }
+                is ProxyMessage.Operations -> {
+                    // If we have no readers, we can ignore updates from storage
+                    if (!this.hasReaders) {
+                        return false;
+                    }
+                    for (op in message.operations) {
+                        if (!this.crdt.applyOperation(op)) {
+                            // If we cannot cleanly apply ops, sync the whole model.
+                            this.clearSynchronized();
+                            return this.requestSynchronization();
+                        }
+                        if (!this.synchronized) {
+                            // If we didn't think we were synchronized but the operation applied cleanly,
+                            // then actually we were synchronized after all. Tell the handle that.
+                            this.setSynchronized();
+                        }
+                        this.notifyUpdate(op);
+                    }
+                }
+                is ProxyMessage.SyncRequest -> {
+                    // storage wants our latest state
+                    val modelUpdate = ProxyMessage.ModelUpdate<Data, Op, T>(this.crdt.data, null)
+                    this.store.onProxyMessage(modelUpdate);
+                }
+                else -> {
```

Comment:
No need for the curly braces here. This could just be:

```
else -> throw CrdtException(...)
```

If you need to line break:

```
else -> throw CrdtException(
    "....."
)
```